web/views.py

```diff
@@ -200,18 +207,100 @@ def form_valid(self, form):
-        # TODO if we need to filter by study demographics vs user demographics
-        # or by if they've taken the study before this is the spot
-        return super().get_queryset().filter(state="active", public=True).order_by("?")
+        user = self.request.user
+        qs = super().get_queryset().filter(state="active", public=True)
+        page = self.request.GET.get("page", 1)
+
+        # Form will return different value depending if the user is logged in.
+        form = self.search_form()
+
+        # values from form
+        search_value = form.data.get("search", "")
+        children_value = form.data.get("children", "")
+        show_experiments_already_done_value = (
+            form.data.get("show_experiments_already_done") == "on"
+        )
+
+        if search_value:
+            qs = qs.filter(name__icontains=search_value)
+
+        # convert to list as it's no longer being treated as a queryset
+        studies = list(qs)
+
+        if children_value:
+            if user.is_anonymous:
+                # when user is anonymous, children value is an age range.
+                age_range = [int(c) for c in children_value.split(",")]
+                studies = [
+                    s for s in studies if age_range_eligibility_for_study(age_range, s)
+                ]
+            else:
+                # when user is authenticated, children value is a child pk
+                child = Child.objects.get(pk=children_value)
+
+                if not show_experiments_already_done_value:
+                    studies = self.completed_consent_frame(studies, child)
+
+            studies = [s for s in studies if get_child_eligibility_for_study(child, s)]
+
+        self.child_eligibility(studies)
+
+        studies.sort(key=self.sort_fn())
+
+        return self.paginated_queryset(studies, page)
+
+    def get_context_data(self, **kwargs):
+        context = super().get_context_data(**kwargs)
+        context["search_form"] = self.search_form()
+        return context
+
+    def completed_consent_frame(self, studies, child):
+        return [
+            r.study
+            for r in Response.objects.filter(
+                study__in=studies, child=child, completed_consent_frame=True,
+            ).distinct("study_id")
+        ]
+
+    def child_eligibility(self, studies):
+        user = self.request.user
+
+        if user.is_authenticated:
+            children = user.children.filter(deleted=False)
+
+            # add eligible children to study object
+            for study in studies:
+                study.eligible_children = create_string_listing_children(
+                    [c for c in children if get_child_eligibility_for_study(c, study)]
+                )
+
+    def search_form(self):
+        if self.request.POST:
+            form = self.form_class(self.request.POST, user=self.request.user)
+            if form.is_valid():
+                return form
+
+        return self.form_class(user=self.request.user)
```

Comment:
I did some digging around and I found [FormMixin](https://docs.djangoproject.com/en/3.2/ref/class-based-views/mixins-editing/#formmixin).  This might be a better pattern to follow and it looks like it's in other views.  Should I take the time to do this refactor?

Reply:
Upon reflection, I think I'd like to do this refactor, if you're willing.

Reply:
Sounds good to me if you're up for it! I wouldn't count on my feeling of "this is odd" - a lot of Django seems odd to me still - but if there's a pattern that makes more sense to you too then let's go for it. Sorry for the delay.

Reply:
I wasn't aware of these mixins.  This is a good opportunity to dig in and learn one of them.